network/v1alpha1/types_dnsnameresolver.go

```diff
@@ -0,0 +1,129 @@
+	Name string `json:"name"`
+}
+
+// DNSNameResolverStatus defines the observed status of DNSNameResolver.
+type DNSNameResolverStatus struct {
+	// resolvedNames contains a list of matching DNS names and their corresponding IP addresses
+	// along with their TTL and last DNS lookup times.
+	// +listType=map
+	// +listMapKey=dnsName
+	// +patchMergeKey=dnsName
+	// +patchStrategy=merge
+	// +optional
+	ResolvedNames []DNSNameResolverResolvedName `json:"resolvedNames,omitempty" patchStrategy:"merge" patchMergeKey:"dnsName"`
+}
+
+// DNSNameResolverResolvedName describes the details of a resolved DNS name.
+type DNSNameResolverResolvedName struct {
+	// conditions provide information about the state of the DNS name.
+	// Known .status.conditions.type is: "Degraded"
+	// +optional
+	// +listType=map
+	// +listMapKey=type
+	Conditions []metav1.Condition `json:"conditions,omitempty"`
+	// dnsName is the resolved DNS name matching the name field of DNSNameResolverSpec. This field can
+	// store both regular and wildcard DNS names which match the spec.name field. When the spec.name
+	// field contains a regular DNS name, this field will store the same regular DNS name after it is
+	// successfully resolved. When the spec.name field contains a wildcard DNS name, each resolvedName.dnsName
+	// will store the regular DNS names which match the wildcard DNS name and have been successfully resolved.
+	// If the wildcard DNS name can also be successfully resolved, then this field will store the wildcard
+	// DNS name as well.
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Pattern=^(\*\.)?([A-Za-z0-9]([-A-Za-z0-9]*[A-Za-z0-9])?\.)+[A-Za-z0-9]([-A-Za-z0-9]*[A-Za-z0-9])?\.$
+	// +kubebuilder:validation:MaxLength=254
+	DNSName string `json:"dnsName"`
+	// resolvedAddresses gives the list of associated IP addresses and their corresponding TTLs and last
+	// lookup times for the dnsName.
+	// +kubebuilder:validation:Required
+	// +listType=map
+	// +listMapKey=ip
+	ResolvedAddresses []DNSNameResolverResolvedAddress `json:"resolvedAddresses"`
+	// resolutionFailures keeps the count of how many consecutive times the DNS resolution failed
+	// for the dnsName. If the DNS resolution succeeds then the field will be set to zero. Upon
+	// every failure, the value of the field will be incremented by one. The details about the DNS
+	// name will be removed, if the value of resolutionFailures reaches 5 and the TTL of all the
+	// associated IP addresses have expired.
+	ResolutionFailures int32 `json:"resolutionFailures,omitempty"`
+}
+
+type DNSNameResolverResolvedAddress struct {
+	// ip is an IP address associated with the dnsName. The validity of the IP address expires after
+	// lastLookupTime + ttlSeconds. To refresh the information a DNS lookup will be performed on the
+	// expiration of the IP address's validity. If the information is not refreshed then it will be
+	// removed with a grace period after the expiration of the IP address's validity.
```

Comment:
How does this grace period interact with `resolutionFailures` count?

Reply:
The grace period is used to wait for that time period before removing the IP address whose TTL has expired. There's actually no direct interaction with the `resolutionFailures` count. The `resolutionFailures` field counts the number of consecutive resolution failures and if the count reaches to 5 and all the associated IP addresses have expired, then the information about the DNS name will be removed from the `resolvedNames` list.

The grace period was suggested so as to account for the delay in DNS lookups and update of the underlying ACL rules. If the IP address is removed immediately on expiration of the TTL and if after DNS resolution on TTL expiration we get the same IP address, then we will be spending time in removing and adding the same IP address from the underlying ACL rules in all the nodes. Additionally, the applications pods may also experience disconnection due to the updates to the ACL rules. To avoid this the IP addresses are removed after a grace period.